[FLINK-36648] Bump Flink version to Flink 2.0.0

lvyanquan opened the pull request:
Bump Flink version to Flink 2.0-preview1.
This is required as flink-example module and flink-e2e-to-e2e-tests need to use Kafka connector for testing, and some other  project like Apache Paimon also rely on Kafka connector in test.

This pr is based on https://github.com/apache/flink-connector-kafka/pull/139 to remove many deprecated classes.
And as there is a bug in SinkTestSuiteBase/SourceTestSuiteBase in Flink 2.0-preview1 as https://issues.apache.org/jira/browse/FLINK-36845 described, temporarily commented relevant test as this will need to wait for a new release of Flink(Or we can directly cover these classes).

lvyanquan commented:
> Bump Flink version to Flink 2.0-preview1. This is required as [flink-example](https://issues.apache.org/jira/browse/FLINK-example) module and [flink-e2e-to-e2e-tests](https://issues.apache.org/jira/browse/FLINK-e2e-to-e2e-tests) need to use Kafka connector for testing, and some other project like Apache Paimon also rely on Kafka connector in test.
> 
> This pr is based on #139 to remove many deprecated classes. And as there is a bug in SinkTestSuiteBase/SourceTestSuiteBase in Flink 2.0-preview1 as https://issues.apache.org/jira/browse/FLINK-36845 described, temporarily commented relevant tests out as this will need to wait for a new release of Flink(Or we can directly cover these classes).

leonardBang commented:
@PatrickRen Would you like to review this PR when you have time?

PatrickRen commented on flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaShortRetentionTestBase.java:
```diff
@@ -128,67 +118,6 @@ public static void shutDownServices() throws Exception {
      */
     private static boolean stopProducer = false;
 
-    public void runAutoOffsetResetTest() throws Exception {
```
This file is not needed anymore. No one is inheriting it.

PatrickRen reviewed (changes requested):
@lvyanquan Thanks for the awesome contribution! I found more classes that can be removed.

I simply searched keyword `deprecated` in the repo, and there are still some appearances. Could you check them all?

Also I see a lot of empty files under `flink-connector-kafka/archunit-violations`. I think we should remove them as well.

lvyanquan commented:
> Also I see a lot of empty files under flink-connector-kafka/archunit-violations. I think we should remove them as well.

Remove some empty files, and remain some files which were referenced in [stored.rules](https://github.com/apache/flink-connector-kafka/blob/main/flink-connector-kafka/archunit-violations/stored.rules).

lvyanquan commented:
Hi, @PatrickRen can you help to review again and help to complete this PR？

lvyanquan commented:
Hi @PatrickRen. I have made adjustments based on your previous suggestion. Can you help confirm if there are any other suggestions?  
Since this PR is likely to conflict with other PRs, it will be troublesome to rebase multiple times in the future. and @AHeise,  I've addressed the comments in https://github.com/apache/flink-connector-kafka/pull/139 too, do you have time to help review the remaining parts？

gyfora commented on .github/workflows/push_pr.yml:
```diff
@@ -28,16 +28,9 @@ jobs:
-        flink: [ 1.20.0 ]
-        jdk: [ '8, 11, 17, 21' ]
+        flink: [ 2.0-preview1 ]
+        jdk: [ '11, 17, 21' ]
```
Since Flink 2.0 is built on Java 17 can we simply build this there as well? Then we can remove Java 11 from the CI matrix

prshnt replied:
Also Flink is out of the preview version now right.

lvyanquan replied:
Done.

lvyanquan commented:
Rebase and rearrange to make commit message more clearly.

lvyanquan commented:
Hi, @PatrickRen @leonardBang @gyfora  could you take a look at this as Flink 2.0.0 was released recently.

mlegore commented:
I hope this can be released soon! Would love to move to 2.0 for our project

leonardBang reviewed (approved):
Thanks @lvyanquan for the great work, LGTM

rmetzger commented on flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/dynamic/source/DynamicKafkaSourceITTest.java:
```diff
@@ -805,30 +801,31 @@ private KafkaStream getKafkaStream(
+    //                        twoKafkas.getContainer().getKafka1(),
+    //                        Collections.emptyList());
+    //    }
 }
```
Why is this test commented out. I fear we are losing test coverage by this

lvyanquan replied:
Thanks for pointing out this, and I've removed the comment.

This was added because there is a bug in SinkTestSuiteBase/SourceTestSuiteBase in Flink 2.0-preview1 as https://issues.apache.org/jira/browse/FLINK-36845 described, and I commented out some related cases to pass CI, but this operation was not need now.

prshnt commented on .github/workflows/weekly.yml:
```diff
@@ -30,7 +30,7 @@ jobs:
       matrix:
         flink_branches: [{
-          flink: 1.20-SNAPSHOT,
+          flink: 2.1-SNAPSHOT,
```
what do you think about using the main branch with the stable release of Flink and not the snapshot?

lvyanquan replied:
I tend to add both snapshot and stable versions to ensure compatibility, and this runs weekly with minimal resource overhead.

prshnt commented:
Thanks @lvyanquan for the PR and taking care of this.

leonardBang commented:
CI passed, merging...

FranMorilloAWS commented:
When is this releasing?
